prov/rxm, verbs: fixes for cmap and EQ

a-ilango opened the pull request:
This PR refactors cmap and fixes serialization issues in verbs EQ, rxm AV and atomics progress.

@swelch please take a look at patches 4 and 5 that modifies auto-progress for atomics.

shefty commented on prov/rxm/src/rxm_cq.c:
```diff
@@ -1382,7 +1382,17 @@ void rxm_ep_do_progress(struct util_ep *util_ep)
+		 * may have been configured with. cmap->acquire would be no-op
+		 * in case rxm data progress is manual */
+		if (rxm_ep->msg_info->domain_attr->threading != FI_THREAD_SAFE)
+			rxm_ep->cmap->acquire(&rxm_ep->cmap->lock);
```
It is really not clear why we need to cmap lock here, particularly, since we're only reading the CQ (call to another provider) and not actually doing any processing while holding the lock.  So, not only does this look like the wrong lock, but it also looks unnecessary.

a-ilango replied:
The atomics auto-progress thread would read MSG CQ as part of fi_trywait. In case we have configured the MSG provider with a threading level other than FI_THREAD_SAFE we need to serialize access to MSG CQ. We're just re-using cmap->lock to do that. Would defining a new lock for this purpose make things clear?

shefty replied:
A new lock would help.  Locking the cmap doesn't look right.  Could the ep lock be used instead?  Or maybe you need a per rxm ep msg_resource_lock?

Part of the confusion is that here you're using the cmap lock to protect access to a resource associated with an EP.  But above (see my other comment), you use the EP lock to protect access to cmap data.  Something looks off.

a-ilango replied:
Over time, the use case of cmap->lock expanded to protecting MSG provider resources as well. I think there was a change added last year which used cmap->lock to guarantee single threaded acccess of libibverbs resources to improve performance. There's a libibverbs env variable which when set would remove all locks in the library.

ep->lock can't be used because it can be no-op when threading level is not FI_THREAD_SAFE and we still have a progress thread in rxm. We could have a msg_resource_lock.

shefty replied:
I would go with msg_resource_lock, associated it with the ep, and remove the cmap->lock if that works.

a-ilango replied:
I have one concern with defining a new msg_resource_lock. We would need to ensure correct lock ordering in case we have to acquire both cmap->lock and msg_resource_lock. I can see this happening when we close a msg ep that failed a simultaneous connection attempt.

As an alternative, would it help if we rename cmap_lock to cm/progress_thread_lock or something like that?

shefty replied:
I was suggesting removing the lock from struct rxm_cmap.  Use an EP related lock (new msg_resource_lock) to protect access to it.  There's a 1:1 relationship between an ep and cmap, correct?  Do we technically even need a separate struct for cmap, versus embedding the fields in the EP?  (I'm not suggesting to do this.)

a-ilango commented on prov/rxm/src/rxm_av.c:
```diff
@@ -83,14 +79,21 @@ rxm_av_insert_cmap(struct fid_av *av_fid, const void *addr, size_t count,
 				continue;
+
+			if (rxm_ep->util_ep.domain->data_progress == FI_PROGRESS_MANUAL)
+				ofi_ep_lock_acquire(&rxm_ep->util_ep);
```
cmap->lock would be NULL in case of manual progress. So we need av lock and ep lock to serialize access to av->ep_list and cmap (an EP resource) respectively.

shefty replied:
If you're protecting access to the cmap, then we should either always use the cmap lock or the ep lock.  Don't change locks based on progress.  It sounds like the issue here was setting cmap lock to null, when it shouldn't be.

a-ilango replied:
ep->lock can be NULL depending on threading level requested by the app. That leaves us with having cmap->lock available always which means in case of manual progress, we are incurring additional cost of acquiring and releasing locks unnecessarily.

a-ilango commented on prov/rxm/src/rxm_av.c:
```diff
@@ -37,29 +37,24 @@ static int rxm_av_remove(struct fid_av *av_fid, fi_addr_t *fi_addr,
-			}
-		}
+		if (rxm_ep->util_ep.domain->data_progress == FI_PROGRESS_MANUAL)
+			ofi_ep_lock_acquire(&rxm_ep->util_ep);
```
May be this optimization is not absolutely needed. I guess until we have data that shows significant performance difference b/w the two we can't say for sure. I can skip this check for now to keep things simple.

swelch commented:
@a-ilango - With commit a2464f709fbe92083c34d2250829a6ea2257a06d I'm seeing a lock issue with an application when using either data FI_PROGRESS_AUTO/FI_PROGRESS_MANUAL with FI_ATOMIC capability. I'm not seeing issues with manual progress and MPI. I will send you directly a back trace of the deadlock I'm seeing.

a-ilango commented:
@swelch I've fixed the lock issue that you're seeing. Please re-test.

Updated the PR with a commit that adds an AV unit test to check av insert at various stages.

swelch commented:
> @swelch I've fixed the lock issue that you're seeing. Please re-test.
> 
> Updated the PR with a commit that adds an AV unit test to check av insert at various stages.

@a-ilango - This fixes the locking issue I was seeing running with this PR. Thanks!

shefty commented:
@a-ilango - feel free to merge if this is working.  Follow on patches can address any cleanup.

a-ilango commented:
> feel free to merge if this is working. Follow on patches can address any cleanup.

Yes, it would be good to do the refactor in a follow up change. I'll merge this once I fix the CI errors.
